// @sap-cloud-sdk/core/dist/connectivity/scp-cf/connectivity-service.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.proxyHostAndPort = exports.addProxyConfigurationOnPrem = exports.addProxyConfiguration = void 0;
var util_1 = require("@sap-cloud-sdk/util");
var protocol_1 = require("./protocol");
var environment_accessor_1 = require("./environment-accessor");
var token_accessor_1 = require("./token-accessor");
var jwt_1 = require("./jwt");
var logger = util_1.createLogger({
    package: 'core',
    messageContext: 'connectivity-service'
});
/**
 * Given a destination and a JWT (required for subscriber destinations), this function will add a proxy configuration to a destination.
 * See also [[ProxyConfiguration]].
 *
 * This function will reject if no connectivity service is bound, no XSUAA service with plan application is bound or the client credentials grant with the XSUAA service fails.
 *
 * @Deprecated Since v1.16.0. Use [[addProxyConfigurationOnPrem]] instead.
 * @param destination - The destination to which the proxy configuration should be added.
 * @param jwt - The JWT of the current user.
 * @returns A promise resolving to the destination with the added proxy configuration.
 */
function addProxyConfiguration(destination, jwt) {
    return Promise.resolve()
        .then(function () { return proxyHostAndPort(); })
        .then(function (hostAndPort) {
        return addHeaders(hostAndPort, destination.authentication, jwt);
    })
        .then(function (proxyConfiguration) { return (__assign(__assign({}, destination), { proxyConfiguration: proxyConfiguration })); });
}
exports.addProxyConfiguration = addProxyConfiguration;
// TODO: remove string argument in v2.0
function addProxyConfigurationOnPrem(destination, jwt) {
    var jwtPair = typeof jwt === 'string' ? { encoded: jwt, decoded: jwt_1.decodeJwt(jwt) } : jwt;
    if (destination.authentication === 'PrincipalPropagation' &&
        !jwt_1.isUserToken(jwtPair)) {
        throw new Error('For principal propagation a user JWT is needed.');
    }
    return addProxyConfiguration(destination, jwtPair === null || jwtPair === void 0 ? void 0 : jwtPair.encoded);
}
exports.addProxyConfigurationOnPrem = addProxyConfigurationOnPrem;
function proxyHostAndPort() {
    var service = readConnectivityServiceBinding();
    return {
        host: service.credentials.onpremise_proxy_host,
        port: service.credentials.onpremise_proxy_http_port ||
            service.credentials.onpremise_proxy_port,
        protocol: protocol_1.Protocol.HTTP
    };
}
exports.proxyHostAndPort = proxyHostAndPort;
function readConnectivityServiceBinding() {
    var serviceBindings = environment_accessor_1.EnvironmentAccessor.getServiceList('connectivity');
    if (!serviceBindings.length) {
        throw new Error('No binding to a connectivity service found! Please make sure to bind an instance of the connectivity service to your app if you want to connect to on-premise destinations.');
    }
    return serviceBindings[0];
}
function addHeaders(hostAndPort, authenticationType, jwt) {
    var connServiceBinding = readConnectivityServiceBinding();
    return Promise.resolve()
        .then(function () { return proxyAuthorizationHeader(connServiceBinding, jwt); })
        .then(function (proxyAuthHeader) { return (__assign(__assign({}, proxyAuthHeader), sapConnectivityAuthenticationHeader(authenticationType, jwt))); })
        .then(function (headers) {
        return (__assign(__assign({}, hostAndPort), { headers: headers }));
    });
}
function proxyAuthorizationHeader(connectivityServiceBinding, userJwt) {
    return token_accessor_1.serviceToken(connectivityServiceBinding, { userJwt: userJwt })
        .then(function (token) { return ({
        'Proxy-Authorization': "Bearer " + token
    }); })
        .catch(function (error) {
        throw new util_1.ErrorWithCause('Failed to add proxy authorization header - client credentials grant failed!', error);
    });
}
function sapConnectivityAuthenticationHeader(authenticationType, jwt) {
    if (authenticationType === 'PrincipalPropagation') {
        if (jwt) {
            return {
                'SAP-Connectivity-Authentication': "Bearer " + jwt
            };
        }
        throw new Error("Unable to create \"SAP-Connectivity-Authentication\" header: no JWT found on the current request.\n     Connecting to on-premise systems via principle propagation is not possible.");
    }
    if (authenticationType === 'BasicAuthentication') {
        logger.warn('You are connecting to an On-Premise system using basic authentication. For productive usage Principal propagation is recommended.');
    }
    return {};
}
